crates/circuit/src/dag_node.rs

```diff
@@ -0,0 +1,329 @@
+                let module = PyModule::import_bound(py, "qiskit.dagcircuit.dagnode")?;
+                Ok(module.getattr("_semantic_eq")?.unbind())
+            })
+            .as_ref()
+            .map_err(|e| e.clone_ref(py))?;
+        semantic_eq_fn.bind(py).call1(args)
+    }
+}
+
+/// Object to represent an Instruction at a node in the DAGCircuit.
+#[pyclass(module = "qiskit._accelerate.circuit", extends=DAGNode)]
+pub struct DAGOpNode {
+    pub instruction: CircuitInstruction,
+    #[pyo3(get)]
+    pub sort_key: PyObject,
+}
+
+type DAGOpNodeState = (((PyObject, Py<PyTuple>, Py<PyTuple>), PyObject), isize);
+
+#[pymethods]
+impl DAGOpNode {
+    #[new]
+    fn new(
+        py: Python,
+        op: PyObject,
+        qargs: Option<&Bound<PySequence>>,
+        cargs: Option<&Bound<PySequence>>,
+        dag: Option<&Bound<PyAny>>,
+    ) -> PyResult<(Self, DAGNode)> {
+        let qargs =
+            qargs.map_or_else(|| Ok(PyTuple::empty_bound(py)), PySequenceMethods::to_tuple)?;
+        let cargs =
+            cargs.map_or_else(|| Ok(PyTuple::empty_bound(py)), PySequenceMethods::to_tuple)?;
+
+        let sort_key = match dag {
+            Some(dag) => {
+                let cache = dag
+                    .getattr(intern!(py, "_key_cache"))?
+                    .downcast_into_exact::<PyDict>()?;
+                let cache_key = PyTuple::new_bound(py, [&qargs, &cargs]);
+                match cache.get_item(&cache_key)? {
+                    Some(key) => key,
+                    None => {
+                        let indices: PyResult<Vec<_>> = qargs
+                            .iter()
+                            .chain(cargs.iter())
+                            .map(|bit| {
+                                dag.call_method1(intern!(py, "find_bit"), (bit,))?
+                                    .getattr(intern!(py, "index"))
+                            })
+                            .collect();
+                        let index_strs: Vec<_> =
+                            indices?.into_iter().map(|i| format!("{:04}", i)).collect();
+                        let key = PyString::new_bound(py, index_strs.join(",").as_str());
+                        cache.set_item(&cache_key, &key)?;
+                        key.into_any()
+                    }
+                }
+            }
+            None => qargs.str()?.into_any(),
+        };
+
+        Ok((
+            DAGOpNode {
+                instruction: CircuitInstruction {
+                    operation: op,
+                    qubits: qargs.unbind(),
+                    clbits: cargs.unbind(),
+                },
+                sort_key: sort_key.unbind(),
+            },
+            DAGNode { _node_id: -1 },
+        ))
+    }
+
+    fn __getstate__(slf: PyRef<Self>, py: Python) -> PyObject {
+        (
+            (slf.instruction.__getstate__(py), &slf.sort_key),
+            (slf.as_ref()._node_id),
+        )
+            .into_py(py)
```

Comment:
How come we're needing to call `__getstate__` manually here?  `pickle` will recurse safely by default, and if we do this manually, we stymie `copy.copy`'s shallow copies.

Reply:
Resolved in 041f3419b06b6f3a8c1a9cb0cc1db0021d357d7c, where I've opted to instead implement `__reduce__`.